### PR TITLE
Update Cirq decomposition to use RZ, RX, CNOT only

### PR DESCRIPTION
**Pull request by Cynocracy**
These gates are preferred for efficiency in our system.

I looked at updating the tests, but they appear to be agnostic to how we decompose and just check correctness, which is great!

Hopefully coverage agrees, will add cases for uncovered lines.

### CODE REVIEW

**Comment by Cynocracy** [comment]
Please forgive any glaring errors, I am only a software middleman for the quantum goodness :) Specifically I don't know enough about gate naming convention to be sure my docs/comments are correct.

File: cirq-ionq/cirq_ionq/ionq_devices.py [hunk 1]
```diff
@@ -103,19 +101,29 @@ def _decompose_single_qubit(self, operation: cirq.Operation) -> cirq.OP_TREE:
...
+        for op in temp.all_operations():
+            if op.gate is None:
+                raise ValueError(
+                    f'Operation {op} is not representable as a gate.'
+                )  # coverage: ignore
+            if type(op.gate) == cirq.ops.phased_x_gate.PhasedXPowGate:
+                gate = cast(cirq.ops.phased_x_gate.PhasedXPowGate, op.gate)
+                yield cirq.ZPowGate(exponent=-gate.phase_exponent).on(op.qubits[0])
+                yield cirq.XPowGate(exponent=gate.exponent).on(op.qubits[0])
+                yield cirq.ZPowGate(exponent=gate.phase_exponent).on(op.qubits[0])
```
Comment (tanujkhattar) [inline cirq-ionq/cirq_ionq/ionq_devices.py:127]: You can do `yield cirq.decompose_once(op)` directly.

File: cirq-ionq/cirq_ionq/ionq_devices.py [hunk 2]
```diff
@@ -103,19 +101,29 @@ def _decompose_single_qubit(self, operation: cirq.Operation) -> cirq.OP_TREE:
...
+        q0, q1 = operation.qubits
+        naive = cirq.two_qubit_matrix_to_operations(q0, q1, mat, allow_partial_czs=False)
+        temp = cirq.Circuit()
+        for op in naive:
+            if type(op.gate) == cirq.ops.common_gates.CZPowGate:
+                temp.append(cirq.H.on(op.qubits[1]))
+                temp.append(cirq.CNOT.on(op.qubits[0], op.qubits[1]))
+                temp.append(cirq.H.on(op.qubits[1]))
+            else:
+                temp.append(op)
```
Comment (tanujkhattar) [inline cirq-ionq/cirq_ionq/ionq_devices.py:115]:
```suggestion
    temp = cirq.map_operations_and_unroll(
        cirq.Circuit(naive),
        lambda op, _: [cirq.H(op.qubits[1]), cirq.CNOT(*op.qubits), cirq.H(op.qubits[1])]
        if type(op.gate) == cirq.CZPowGate
        else op,
    )
```

File: cirq-ionq/cirq_ionq/ionq_devices.py [hunk 3]
```diff
@@ -103,19 +101,29 @@ def _decompose_single_qubit(self, operation: cirq.Operation) -> cirq.OP_TREE:
...
+                raise ValueError(
+                    f'Operation {op} is not representable as a gate.'
+                )  # coverage: ignore
+            if type(op.gate) == cirq.ops.phased_x_gate.PhasedXPowGate:
+                gate = cast(cirq.ops.phased_x_gate.PhasedXPowGate, op.gate)
+                yield cirq.ZPowGate(exponent=-gate.phase_exponent).on(op.qubits[0])
+                yield cirq.XPowGate(exponent=gate.exponent).on(op.qubits[0])
+                yield cirq.ZPowGate(exponent=gate.phase_exponent).on(op.qubits[0])
+            else:
+                yield op
```
Comment (tanujkhattar) [inline cirq-ionq/cirq_ionq/ionq_devices.py:129]:
```suggestion
    temp = cirq.map_operations_and_unroll(
        circuit,
        lambda op, _: cirq.decompose_once(op) if type(op.gate) == cirq.PhasedXPowGate else op,
    )
    yield temp.all_operations()
```

File: cirq-ionq/cirq_ionq/ionq_devices.py (hunk 2 above)
Comment (Cynocracy) [reply cirq-ionq/cirq_ionq/ionq_devices.py:115]: Beautiful suggestions, thank you!!

File: cirq-ionq/cirq_ionq/ionq_devices.py [hunk 4]
```diff
@@ -103,19 +101,19 @@ def _decompose_single_qubit(self, operation: cirq.Operation) -> cirq.OP_TREE:
...
-                yield gate(qubit)
+        q0, q1 = operation.qubits
+        naive = cirq.two_qubit_matrix_to_operations(q0, q1, mat, allow_partial_czs=False)
+        temp = cirq.map_operations_and_unroll(
+            cirq.Circuit(naive),
+            lambda op, _: [cirq.H(op.qubits[1]), cirq.CNOT(*op.qubits), cirq.H(op.qubits[1])]
+            if type(op.gate) == cirq.CZPowGate
+            else op,
+        )
+        cirq.optimizers.merge_single_qubit_gates.merge_single_qubit_gates_into_phased_x_z(temp)
```
Comment (tanujkhattar) [inline cirq-ionq/cirq_ionq/ionq_devices.py:114]:
```suggestion
        cirq.merge_single_qubit_gates_into_phased_x_z(temp)
```

File: cirq-ionq/cirq_ionq/ionq_devices.py [hunk 5]
```diff
@@ -103,19 +101,19 @@ def _decompose_single_qubit(self, operation: cirq.Operation) -> cirq.OP_TREE:
             yield gate(qubit)
 
     def _decompose_two_qubit(self, operation: cirq.Operation) -> cirq.OP_TREE:
-        """Decomposes a two qubit gate into XXPow, YYPow, and ZZPow plus single qubit gates."""
+        """Decomposes a two qubit unitary Rz,Rx,and CNOT."""
```
Comment (tanujkhattar) [inline cirq-ionq/cirq_ionq/ionq_devices.py:104]: This is a pretty big change and I'm surprised that none of the existing tests fail after this change. 

Please add tests which would pass now but would have failed if the change is reverted.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Very soon, the `decompose_operation` API on the device is going to be deprecated (see https://github.com/quantumlib/Cirq/issues/4744), so ideally we should move the compilation logic out of the device class to a standalone transformer. 

But to unblock this particular change, I'm fine with going ahead with it for now and doing the migration to an external transformer in a separate follow-up PR.

File: cirq-ionq/cirq_ionq/ionq_devices.py (hunk 5 above)
Comment (tanujkhattar) [inline cirq-ionq/cirq_ionq/ionq_devices.py:104]:
```suggestion
        """Decomposes a two qubit unitary operation into Rz, Rx, and CNOT."""
```

Comment (tanujkhattar) [reply cirq-ionq/cirq_ionq/ionq_devices.py:104]: The test should also be useful to guard against any potential change to the default decomposition of `PhasedXPowGate` into `Rx` / `Rz` gates.

Comment (Cynocracy) [reply cirq-ionq/cirq_ionq/ionq_devices.py:104]: Commit!

Comment (Cynocracy) [reply cirq-ionq/cirq_ionq/ionq_devices.py]: Makes sense! I added a check that the valid gateset is used, in addition to the unitary equivalence check being done. WDYT?

**Comment by Cynocracy** [comment]
sorry for the slow turnaround :) here you are

**Comment by Cynocracy** [comment]
If you point me in the direction of an example migration to transformer, I can start that PR :)

**Review by tanujkhattar — APPROVED** [review]
LGTM!

For transformers, you can open an issue and I can add some comments there. It'll be easier to track the migration.

**Comment by Cynocracy** [comment]
cut https://github.com/quantumlib/Cirq/issues/4901

Thanks for the review! :)

**Pull request merged**